### PR TITLE
Implement RandomSearch

### PR DESCRIPTION
**Pull request by ablaom**
This is more-or-less the implementation proposed at https://github.com/alan-turing-institute/MLJ.jl/issues/37, with addition that tuples of the form `(hyperparameter_name, sampler)` can be specified in the range vector, where `sampler` is just something implementing `rng`. 

Here's the updated doc-string:

---

    RandomSearch(bounded=Distributions.Uniform,
                 positive_unbounded=Distributions.Gamma,
                 other=Distributions.Normal,
                 rng=Random.GLOBAL_RNG)

Instantiate a random search tuning strategy, for searching over
Cartesian hyperparameter domains, with customizable priors in each
dimenension.

### Supported ranges:

- a single one-dimensional range (`ParamRange` object) `r`

- a pair of the form `(r, d)`, with `r` as above and where `d` is a
  probability vector of the same length as `r.values`, if `r` is a
  `NominalRange`, and is otherwise: (i) any
  `Distributions.UnivariateDistribution` *instance*; or (ii) one of
  the *subtypes* of `Distributions.UnivariateDistribution` listed in
  the table below, for automatic fitting using `Distributions.fit(d,
  r)` (a distribution whose support always lies between `r.lower` and
  `r.upper`.)

- any pair of the form `(field, s)`, where `field` is the (possibly
  nested) name of a field of the model to be tuned, and `s` an
  arbitrary sampler object for that field. This means only that
  `rand(rng, s)` is defined and returns valid values for the field.

- any vector of objects of the above form

distribution types  | for fitting to ranges of this type
--------------------|-----------------------------------
`Arcsine`, `Uniform`, `Biweight`, `Cosine`, `Epanechnikov`, `SymTriangularDist`, `Triweight` | bounded
`Gamma`, `InverseGaussian`, `Poisson` | positive (bounded or unbounded)
`Normal`, `Logistic`, `LogNormal`, `Cauchy`, `Gumbel`, `Laplace`  | any

`ParamRange` objects are constructed using the `range` method.

### Examples:

    range1 = range(model, :hyper1, lower=1, origin=2, unit=1)

    range2 = [(range(model, :hyper1, lower=1, upper=10), Arcsine),
               range(model, :hyper2, lower=2, upper=4),
              (range(model, :hyper2, lower=2, upper=4), Normal(0, 3)),
               range(model, :hyper3, values=[:ball, :tree], [0.3, 0.7])]

    # uniform sampling of :(atom.λ) from [0, 1] without defining a NumericRange:
    struct MySampler end
    Base.rand(rng::AbstractRNG, ::MySampler) = rand(rng)
    range3 = (:(atom.λ), MySampler(), range1)

### Algorithm

Models for evaulation are generated by sampling each range `r` using
`rand(rng, s)` where, `s = sampler(r, d)`. See `sampler` for details. If `d`
is not specified, then sampling is uniform (with replacement) in the
case of a `NominalRange`, and is otherwise given by the defaults
specified by the tuning strategy parameters `bounded`,
`positive_unbounded`, and `other`, depending on the `NumericRange`
type.

See also [`TunedModel`](@ref), [`range`](@ref), [`sampler`](@ref).

### CODE REVIEW

**Comment by ablaom** [comment]
@tlienart A superficial look over would be great, only if you have time.

File: README.md [hunk 1]
```diff
@@ -232,6 +232,8 @@ In setting up a tuning task, the user constructs an instance of the
 
 ### Implementation requirements for new tuning strategies
 
+As sample implemenations, see [/src/strategies/](/src/strategies)
```
Comment (tlienart) [inline README.md:235]: implementation

File: src/range_methods.jl [hunk 2]
```diff
@@ -0,0 +1,145 @@
+## BOUNDEDNESS TRAIT
+
+# For random search and perhaps elsewhere, we need a variation on the
+# built-in boundedness notions:
+abstract type PositiveUnbounded <: Unbounded end
```
Comment (tlienart) [inline src/range_methods.jl:5]: Why a d difference between PositiveUnbounded and RightUnbounded ?

File: src/strategies/random_search.jl [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+Cartesian hyperparameter domains, with customizable priors in each
+dimension.
+
+### Supported ranges
+
+- a single one-dimensional range (`ParamRange` object) `r`
+
+- a pair of the form `(r, d)`, with `r` as above and where `d` is a
+  probability vector of the same length as `r.values`, if `r` is a
+  `NominalRange`, and is otherwise: (i) any
```
Comment (tlienart) [inline src/strategies/random_search.jl:19]: formatting maybe put each of the option in separate paragraph

File: src/range_methods.jl (hunk 2 above)
Comment (ablaom) [reply src/range_methods.jl:5]: Well, initially I thought of translating distributions when fitting to a right unbounded range with `lower` not equal to zero. But the sparsity of use-cases I could imagine did not seem to justify the much larger extra implementation complexity. So now, if you are positive unbounded range `r`, then the value of `r.lower` is used only to truncate, not to fit. To fit we just look at `unit` and `origin` to fit the (two-parameter) distribution specified (and then truncate). Most of the time `r.lower=0` and you are probably specifying a positive distribution type to fit anyway. 

At the end of day, if you don't like the automatically fit distribution, you can specify your own.

**Comment by ablaom** [comment]
Many thanks @tlienart .

**Pull request merged**